## L05 — Ćwiczenia laboratoryjne
#### Temat: Pandas — Series, DataFrame, eksploracja danych

In [1]:
import pandas as pd
import numpy as np
# Series z listy — domyślny indeks (0, 1, 2, ...)
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny)

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0
dtype: float64


In [3]:
#Zadanie 1a
pensje = pd.Series({
    'Anna':  5500,
    'Jan':    7200,
    'Ewa':   4800,
    'Marek':    9100,
    'Kasia':     6300
})


In [11]:
#Zadanie 1b
#1.
osoba_max = pensje.idxmax()
print(f"Osoba zarabiająca najwięcej: {osoba_max}")
#2.
srednia_pensja = pensje.mean()
print(f"Średnia pensja: {srednia_pensja}")
#3.
maska = pensje > srednia_pensja
osoby_powyzej_sredniej = pensje[maska].index
print(f"Osoby zarabiające powyżej średniej: {list(osoby_powyzej_sredniej)}")
#4.
maska = pensje < 6000
osoby_ponizej_sredniej = pensje[maska].index
print(f"Ilość osób zarabiających poniżej średniej: {len(osoby_ponizej_sredniej)}")


Osoba zarabiająca najwięcej: Marek
Średnia pensja: 6580.0
Osoby zarabiające powyżej średniej: ['Jan', 'Marek']
Ilość osób zarabiających poniżej średniej: 2


In [23]:
#Zadanie 1c
#1.
print(f"Pensja po podwyżce 10%: ")
po_podwyzce_10_procent = pensje * 1.10
print(po_podwyzce_10_procent.to_string())
#2.
print(f"\nPensja po premii 500zł: ")
po_premii_500 = pensje + 500
print(po_premii_500.to_string())
#3.
print(f"\nNowa pensja: ")
nowa_pensja = (pensje * 1.10) + 500
print(nowa_pensja.to_string())
#4.
print(f"\nRanking nowych pensji od najwyższej: ")
print(nowa_pensja.sort_values(ascending=False).to_string())

Pensja po podwyżce 10%: 
Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0

Pensja po premii 500zł: 
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800

Nowa pensja: 
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0

Ranking nowych pensji od najwyższej: 
Marek    10510.0
Jan       8420.0
Kasia     7430.0
Anna      6550.0
Ewa       5780.0


#### Ćwiczenie 2: DataFrame — tworzenie i atrybuty (20 min)

In [ ]:
# Dane o produktach sklepu internetowego
sklep = pd.DataFrame({
'produkt': ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
'cena': [3500, 1800, 2500, 350, 2200],
'stan_magazynowy': [45, 120, 200, 500, 75],
'kategoria': ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']
})

In [30]:
#Zadanie 2a
# 1. Wyświetl kształt tabeli (shape)
print(f"Kształt tabeli:")
print(sklep.shape)
# 2. Wyświetl nazwy kolumn (columns)
print(f"\nNazwy kolumn:")
print(sklep.columns)
# 3. Wyświetl typy danych w każdej kolumnie (dtypes)
print(f"\nTypy danych:")
print(sklep.dtypes)
# 4. Wyświetl pierwsze 3 wiersze (head)
print(f"\nPierwsze 3 wiersze:")
print(sklep.head)

Kształt tabeli:
(5, 4)

Nazwy kolumn:
Index(['produkt', 'cena', 'stan_magazynowy', 'kategoria'], dtype='str')

Typy danych:
produkt              str
cena               int64
stan_magazynowy    int64
kategoria            str
dtype: object

Pierwsze 3 wiersze:
<bound method NDFrame.head of      produkt  cena  stan_magazynowy  kategoria
0     Laptop  3500               45  Komputery
1     Tablet  1800              120    Mobilne
2   Smartfon  2500              200    Mobilne
3  Słuchawki   350              500  Akcesoria
4    Monitor  2200               75  Komputery>


In [34]:
#Zadanie 2b
# 1. Dodaj kolumnę 'wartosc_magazynu' = cena * stan_magazynowy
sklep['wartosc_magazynu'] = (sklep['cena'] * sklep['stan_magazynowy'])
# 2. Dodaj kolumnę 'cena_z_vat' = cena * 1.23
sklep['cena_z_vat'] = (sklep['cena'] * 1.23)
# 3. Wyświetl cały DataFrame z nowymi kolumnami
print(sklep)

     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_vat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0


In [35]:
# Wczytaj dataset tips
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url)

In [44]:
#Zadanie 2c
# 1. Wyświetl ile wierszy i kolumn ma dataset (shape)
print("Ile wierszy i kolumn ma dataset: ")
print(tips.shape)
# 2. Wyświetl ostatnie 5 wierszy (tail)
print("\nOstatnie 5 wierszy: ")
print(tips.tail())
# 3. Wyświetl 3 losowe wiersze (sample)
print("\n3 losowe wiersze: ")
print(tips.sample(3))
# 4. Wyświetl typy danych w każdej kolumnie (dtypes)
print("\nTypy danych w każdej kolumnie: ")
print(tips.dtypes)


Ile wierszy i kolumn ma dataset: 
(244, 7)

Ostatnie 5 wierszy: 
     total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

3 losowe wiersze: 
    total_bill   tip   sex smoker   day    time  size
77       27.20  4.00  Male     No  Thur   Lunch     4
88       24.71  5.85  Male     No  Thur   Lunch     2
31       18.35  2.50  Male     No   Sat  Dinner     4

Typy danych w każdej kolumnie: 
total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object


#### Ćwiczenie 3: Eksploracja datasetu — samodzielna analiza (30 min)

In [45]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url)

In [50]:
#Zadanie 3a
# 1. Wyświetl head(10)
print(penguins.head(10))

  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       36

In [51]:
# 2. Wyświetl shape
print(penguins.shape)

(344, 7)


In [49]:
# 3. Uruchom info() — ile brakujących wartości widzisz?
penguins.info()

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    str    
 1   island             344 non-null    str    
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    str    
dtypes: float64(4), str(3)
memory usage: 18.9 KB


In [52]:
#Zadanie 3b
# 1. Uruchom describe() — co mówią ci średnie i odchylenia?
penguins.describe()

,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


In [53]:
# 2. Uruchom describe(include='all') — co dodają kolumny tekstowe?
penguins.describe(include='all')

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
count,344,344,342.000000,342.000000,342.000000,342.000000,333
unique,3,3,NaN,NaN,NaN,NaN,2
top,Adelie,Biscoe,NaN,NaN,NaN,NaN,MALE
freq,152,168,NaN,NaN,NaN,NaN,168
mean,NaN,NaN,43.921930,17.151170,200.915205,4201.754386,NaN
std,NaN,NaN,5.459584,1.974793,14.061714,801.954536,NaN
min,NaN,NaN,32.100000,13.100000,172.000000,2700.000000,NaN
25%,NaN,NaN,39.225000,15.600000,190.000000,3550.000000,NaN
50%,NaN,NaN,44.450000,17.300000,197.000000,4050.000000,NaN
75%,NaN,NaN,48.500000,18.700000,213.000000,4750.000000,NaN


In [56]:
# 3. Jaka jest mediana masy ciała (body_mass_g)?
mediana_masy = penguins['body_mass_g'].median()
print(f"Mediana masy ciała: {mediana_masy}g.")

Mediana masy ciała: 4050.0g.


In [57]:
#Zadanie 3c
# 1. Ile jest pingwinów każdego gatunku? (value_counts na 'species')
print(penguins['species'].value_counts())

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64


In [58]:
# 2. Ile pingwinów na każdej wyspie? (value_counts na 'island')
print(penguins['island'].value_counts())

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64


In [59]:
# 3. Rozkład płci (value_counts na 'sex') — czy jest zbalansowany?
print(penguins['sex'].value_counts())

sex
MALE      168
FEMALE    165
Name: count, dtype: int64


In [60]:
# 4. Jaki procent to gatunek Adelie? (normalize=True)
print(penguins['species'].value_counts(normalize=True))

species
Adelie       0.441860
Gentoo       0.360465
Chinstrap    0.197674
Name: proportion, dtype: float64


In [61]:
#Zadanie 3d
# 1. Policz brakujące wartości w każdej kolumnie (isna().sum())
print(penguins.isna().sum())

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64


In [62]:
# 2. Jaki procent danych brakuje w kolumnie 'sex'?
procent_brakow_sex = penguins['sex'].isna().sum() / len(penguins) * 100
print(f"Procent brakujących danych w 'sex': {procent_brakow_sex:.2f}%")

Procent brakujących danych w 'sex': 3.20%


1. Dataset zawiera informacje o 344 pingwinach opisanych za pomocą 7 różnych cech (kolumn).

2. Najliczniejszym gatunkiem jest Adelie, który liczy 152 osobniki i stanowi około 44,19% całego zbioru.

3. Średnia masa ciała pingwinów wynosi około 4201,75 g, podczas gdy mediana jest niższa i wynosi dokładnie 4050,0 g.

4. Badane pingwiny zamieszkują 3 wyspy (Biscoe, Dream, Torgersen), z czego najliczniejszą populację (168 osobników) spotkamy na wyspie Biscoe.

5. Najwięcej brakujących informacji występuje w kolumnie 'sex', gdzie brakuje danych dla 11 osobników, co stanowi około 3,20% zbioru.

#### Ćwiczenie 4: Pytania biznesowe + commit (15 min)

In [63]:
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')

In [64]:
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)

In [65]:
# 1. Oblicz średni rachunek (total_bill)
sredni_rachunek = tips['total_bill'].mean()
print(f"1. Średni rachunek: {sredni_rachunek:.2f}")

1. Średni rachunek: 19.79


In [66]:
# 2. Oblicz średni napiwek w procentach (tip_pct)
sredni_napiwek_pct = tips['tip_pct'].mean()
print(f"2. Średni napiwek (%): {sredni_napiwek_pct:.2f}%")

2. Średni napiwek (%): 16.08%


In [71]:
# 3. Wyświetl który dzień ma najwięcej rachunków (value_counts na 'day')
dni_counts = tips['day'].value_counts()
print(f"\nLiczba rachunków w podziale na dni:\n {dni_counts}")
print(f"Najwięcej rachunków jest w: {dni_counts.idxmax()}")


Liczba rachunków w podziale na dni:
 day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64
Najwięcej rachunków jest w: Sat


In [73]:
# 4. Policz ile rachunków jest od palaczy (value_counts na 'smoker')
palacze_counts = tips['smoker'].value_counts()
print(f"\nLiczba rachunków (palacze vs niepalący): {palacze_counts}")


Liczba rachunków (palacze vs niepalący): smoker
No     151
Yes     93
Name: count, dtype: int64


In [75]:
# 5. Wyświetl describe() tylko dla kolumny 'tip' — jaka jest mediana napiwku?
tip_stats = tips['tip'].describe()
mediana_napiwku = tips['tip'].median()
print(f"\nStatystyki dla kolumny 'tip': {tip_stats}")
print(f"Mediana napiwku wynosi: {mediana_napiwku}")


Statystyki dla kolumny 'tip': count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000
Name: tip, dtype: float64
Mediana napiwku wynosi: 2.9


In [ ]:
# 6. Jaki procent rachunków pochodzi z obiadu (Dinner)?
procent_czasu = tips['time'].value_counts(normalize=True) * 100
print(f"\nProcentowy udział porposiłków: {procent_czasu}")
print(f"Z obiadu (Dinner) pochodzi: {procent_czasu['Dinner']:.2f}% rachunków")


6. Procentowy udział porposiłków: time
Dinner    72.131148
Lunch     27.868852
Name: proportion, dtype: float64
Z obiadu (Dinner) pochodzi: 72.13% rachunków


In [78]:
# 7. Ile unikalnych rozmiarów grup jest w danych? (nunique na 'size')
unikalne_rozmiary = tips['size'].nunique()
print(f"\nLiczba unikalnych rozmiarów grup: {unikalne_rozmiary}")


Liczba unikalnych rozmiarów grup: 6
